File: Leetcode/Ways to Make a Fair Array/Ways to Make a Fair Array.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
    //declaration of arrays
    int presumEven[100005];
    int presumOdd[100005];
public:
    //function
    int waysToMakeFair(vector<int>& nums)
    {
        //calculating size of the given vector
        int n = nums.size();
        
        //if vector have only 1 element return 1
        if(nums.size() == 1) return 1;
        
        //if vector have only two elements return 0
        if(nums.size() == 2) return 0;
        
        //copying the value of element at index 0 and 1 in the declared arrays of even and odd
        presumEven[0] = presumEven[1] = nums[0];
        presumOdd[0] = 0;
        presumOdd[1] = nums[1];
        
        //copying elements of odd and even indexes to their respective arrays 
        for(int i = 1; i < n; ++i) 
        {
            if(i % 2 == 1)
            {
                presumOdd[i] = presumOdd[i-1] + nums[i];
                presumEven[i] = presumEven[i-1];
                
            }
            else 
            {
                presumOdd[i] = presumOdd[i-1];
                presumEven[i] += presumEven[i-1] + nums[i];
            }
        }
        
        //declaration of variable ret
        int ret = 0;
        
        //checking if elements of even and odd indexes are same 
        if(presumEven[n-1] - nums[0] == presumOdd[n-1]) 
        {
            ret++;
        }
        
        //loop for removing element
        for(int i = 1; i < n; ++i)
        {
            // removing the current index
            int oddSum = presumOdd[i-1] + (presumEven[n-1] - presumEven[i]);
            int evenSum = presumEven[i-1] + (presumOdd[n-1] - presumOdd[i]);
            
            //checking if sum of od indexed values is equal to the sum of even indexed values 
            if(oddSum == evenSum) 
            {
                ret++;
            }
        }
        
        //returning the number of indices
        return ret;
    }
};

File: Leetcode/Ways to Make a Fair Array/Ways to Make a Fair Array.java
class Solution {
    public int waysToMakeFair(int[] nums) {
        //Finding the size of given array
        int size= nums.length;
        //Declaring the answer
        int ans=0;
        //Now we will have two arrays. The left array is for all the elements that are yet to be processed and removed array is for all the elements that are already processed. We will further divide them based on if they are odd or even, and then we will maintain odd sum and even sum.
        int[] left= new int[2];
        int[] removed= new int[2];
        //Initially adding all the elements to left array
        for(int i=0;i<size;i++){
            left[i%2]+= nums[i];
        }
        //Processing each element
        for(int i=0;i<size;i++){
            left[i%2]-= nums[i];
            //After removing the current element, we are now checking if the even sum is equal to odd sum of all the other elements. Note that after removing the current element, the elements that come after that element will have their index parity changed
            if(left[0]+ removed[1]== left[1]+ removed[0]){
                ans++;
            }
            removed[i%2]+= nums[i];
        }
        return ans;
    }
}
